packages/i18n/src/index.ts

```diff
@@ -137,14 +137,17 @@ export const prefix: <T extends BaseRecordDict, P extends string>(
   return result;
 };
 
-export type BaseTemplateArgs = Record<string, string | number | boolean>;
+type DefaultResolveArgs = string | number | boolean;
+type DefaultResolved = string
+
+export type BaseTemplateArgs<Args = DefaultResolveArgs> = Record<string, Args>;
 
 /**
  * A string branded with arguments needed to resolve the template.
  */
-export type Template<T extends BaseTemplateArgs> = string & { __template_args: T };
+export type Template<T extends BaseTemplateArgs<Args>, Args> = string & { __template_args: T };
```

Comment:
Not sure what you mean by that, but I have removed the Args anyways because it doesn't really make sense to introduce a generic. The reson is that the type of the resolver cannot really be transferred to the template call. See my comment on the `template` function:

> Keep in mind that you may only use arguments of a type that is supported by your template resolver

If the arguments used by the template are not supported by the resolver than I have introduced this sketchy safety machanism:

```
export type ResolveArgs<Value, Args> = Value extends (...args: infer A) => any
  ? A
  : Value extends Template<infer R>
    ? (R extends BaseTemplateArgs<Args> ? [args: R] : [`Translator resolver doesn't fully support the argument types used in this template`])
    : Value extends BaseRecordDict ?
      [] :
      [args?: BaseTemplateArgs<Args>];
```

I also don't like it a lot, is everything else than pretty, but it works reliable and does it's job. The other thing we could have done is to just ignore those entries by returning an empty array, but I think that confuses developers, because there is no way one can conclude out of the fact that the template translation key is not available, that they have used an unsupported argument.